File: prep/coreprep/stack/ParensBalance.java
import java.util.Scanner;

public class ParensBalance {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();
      in.nextLine();

      while (tt-- != 0) {
        String input = in.nextLine().trim();
        Stack<Character> st = new Stack<>(100);
        boolean valid = true;

        for (int i = 0; i < input.length(); i++) {
          char c = input.charAt(i);
          if (c == '(' || c == '[' || c == '{') {
            st.push(c);
          } else if (c == ')' || c == ']' || c == '}') {
            if (st.isEmpty()) {
              valid = false;
              break;
            }

            char d = st.pop();
            if ((c == ')' && d != '(') || (c == ']' && d != '[') ||
                (c == '}' && d != '{')) {
              valid = false;
              break;
            }
          }
        }

        if (!valid || !st.isEmpty()) {
          System.out.println("Invalid string");
        } else {
          System.out.println("Valid string");
        }
      }
    }
  }
}

File: lc/coreprep/topics/stack/parens_balance.py
def main():
    tt = int(input())
    for _ in range(tt):
        s = input().strip()

        st = []
        valid = True
        for c in s:
            if c in ['(', '[', '{']:
                st.append(c)
            elif c in [')', ']', '}']:
                if len(st) == 0:
                    valid = False
                    break
                t = st.pop()
                if (c == ')' and t != '(') or (c == ']' and t != '[') or (c == '}' and t != '{'):
                    valid = False
                    break

        if not valid or len(st) != 0:
            print("Invalid string")
        else:
            print("Valid string")


if __name__ == "__main__":
    main()